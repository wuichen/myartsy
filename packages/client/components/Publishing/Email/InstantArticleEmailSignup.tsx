import * as Fonts from "Assets/Fonts"
import React from "react"
import styled, { StyledFunction } from "styled-components"
import request from "superagent"
import Colors from "../../../Assets/Colors"
import InvertedButton from "../../Buttons/Inverted"
import { borderedInput } from "../../Mixins"
import { EMAIL_REGEX } from "../Constants"

interface EmailProps {
  signupUrl: string
}

interface EmailState {
  value: string
  error: boolean
  disabled: boolean
  message: string
}

interface InputProps {
  isError: boolean
  isReadOnly: boolean
}

export class InstantArticleEmailSignup extends React.Component<
  EmailProps,
  EmailState
> {
  constructor(props) {
    super(props)
    this.state = {
      value: "",
      error: false,
      disabled: false,
      message: "",
    }
  }

  onClick = () => {
    this.setState({ disabled: true })
    if (this.state.value.match(EMAIL_REGEX)) {
      request
        .post(this.props.signupUrl)
        .send({ email: this.state.value })
        .set("accept", "json")
        .end((err, res) => {
          if (err) {
            this.flashMessage("Error. Please try again", true)
          } else {
            this.flashMessage("Thank you!", false)
          }
        })
    } else {
      this.flashMessage("Invalid Email... Please try again", true)
    }
  }

  flashMessage = (message, error) => {
    this.setState({ message, error })
    setTimeout(() => {
      this.setState({ message: "", disabled: false, error: false })
    }, 2000)
  }

  onInputChange = e => {
    this.setState({ value: e.target.value, error: null })
  }

  render() {
    return (
      <InstantArticleEmailContainer>
        <Subtitle>The Daily Editorial Newsletter</Subtitle>
        <Title>The latest art-world news and stories in your inbox</Title>
        <Form>
          <Input
            type="email"
            placeholder="Enter Your Email..."
            onChange={this.onInputChange}
            value={this.state.message || this.state.value}
            readOnly={this.state.message.length > 0}
            isError={this.state.error}
            isReadOnly={this.state.message.length > 0}
          />
          <StyledButton disabled={this.state.disabled} onClick={this.onClick}>
            Subscribe
          </StyledButton>
        </Form>
      </InstantArticleEmailContainer>
    )
  }
}

const input: StyledFunction<InputProps & React.HTMLProps<HTMLInputElement>> =
  styled.input
const Input = input`
  ${borderedInput}
  width: 100%;
  border-width: 1px;
  color: ${props => (props.isError ? Colors.redMedium : "black")};
  ${props => (props.isReadOnly ? Fonts.unica("s16") : "")}
`
const InstantArticleEmailContainer = styled.div`
  display: flex;
  flex-direction: column;
  max-width: 500px;
  padding: 30px 0;
  margin: 0 20px;
  border-top: 1px solid ${Colors.grayRegular};
  border-bottom: 1px solid ${Colors.grayRegular};
`
const Title = styled.div`
  ${Fonts.garamond("s23")};
  margin-bottom: 30px;
`
const Subtitle = styled.div`
  ${Fonts.avantgarde("s11")};
  margin-bottom: 10px;
`
const StyledButton = styled(InvertedButton)`
  border-radius: 2px;
  height: 30px;
  width: 80px;
  margin-left: -100px;
  ${Fonts.avantgarde("s11")};
`
const Form = styled.div`
  display: flex;
`
