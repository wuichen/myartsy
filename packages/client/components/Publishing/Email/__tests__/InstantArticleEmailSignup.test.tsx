import { mount } from "enzyme"
import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import request from "superagent"
import { InstantArticleEmailSignup } from "../InstantArticleEmailSignup"

jest.mock("superagent", () => {
  const end = jest.fn()
  const set = jest.fn().mockReturnValue({ end })
  const send = jest.fn().mockReturnValue({ set })
  const post = jest.fn().mockReturnValue({ send })
  return { post, set, send, end }
})
jest.useFakeTimers()

describe("EmailSignup", () => {
  it("renders an email signup", () => {
    const emailSignup = renderer.create(
      <InstantArticleEmailSignup signupUrl="#" />
    )
    expect(emailSignup).toMatchSnapshot()
  })

  it("submits an email", () => {
    const viewer = mount(<InstantArticleEmailSignup signupUrl="#" />)
    viewer.setState({ value: "foo@goo.net" })
    viewer.find("button").simulate("click")
    expect(request.post).toBeCalled()
    expect(request.send.mock.calls[0][0].email).toEqual("foo@goo.net")
    request.end.mock.calls[0][0]()
    const state = viewer.state() as any
    expect(state.message).toEqual("Thank you!")

    jest.runAllTimers()
    const postTimeoutState = viewer.state() as any
    expect(postTimeoutState.disabled).toBe(false)
    expect(postTimeoutState.message).toEqual("")
  })

  it("handles signup errors", () => {
    const viewer = mount(<InstantArticleEmailSignup signupUrl="#" />)
    viewer.setState({ value: "foo@goo.net" })
    viewer.find("button").simulate("click")
    expect(request.post).toBeCalled()
    request.end.mock.calls[1][0]("Error")
    const state = viewer.state() as any
    expect(state.message).toEqual("Error. Please try again")
    expect(state.error).toBe(true)

    jest.runAllTimers()
    const postTimeoutState = viewer.state() as any
    expect(postTimeoutState.disabled).toBe(false)
    expect(postTimeoutState.error).toBe(false)
    expect(postTimeoutState.message).toEqual("")
  })

  it("validates email addresses", () => {
    const viewer = mount(<InstantArticleEmailSignup signupUrl="#" />)
    viewer.setState({ value: "foo" })
    viewer.find("button").simulate("click")
    const state = viewer.state() as any
    expect(state.message).toEqual("Invalid Email... Please try again")
    expect(state.error).toBe(true)

    jest.runAllTimers()
    const postTimeoutState = viewer.state() as any
    expect(postTimeoutState.disabled).toBe(false)
    expect(postTimeoutState.error).toBe(false)
    expect(postTimeoutState.message).toEqual("")
  })
})
