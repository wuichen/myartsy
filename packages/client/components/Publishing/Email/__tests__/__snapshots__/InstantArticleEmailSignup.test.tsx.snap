// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailSignup renders an email signup 1`] = `
.c4 {
  padding: 10px;
  box-shadow: none;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 10px;
  resize: none;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 1px solid #e5e5e5;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  width: 100%;
  border-width: 1px;
  color: black;
}

.c4::-webkit-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c4::-moz-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c4:-ms-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c4::placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c4:hover,
.c4:focus,
.c4.focused {
  border-color: #6e1fff;
  outline: 0;
}

.c4:disabled {
  border: 2px dotted #e5e5e5;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 500px;
  padding: 30px 0;
  margin: 0 20px;
  border-top: 1px solid #e5e5e5;
  border-bottom: 1px solid #e5e5e5;
}

.c2 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 23px;
  line-height: 1.5em;
  -webkit-font-smoothing: antialiased;
  margin-bottom: 30px;
}

.c1 {
  font-family: 'ITC Avant Garde Gothic W04','AvantGardeGothicITCW01D 731075',AvantGardeGothicITCW01Dm,Helvetica,sans-serif;
  -webkit-font-smoothing: antialiased;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-size: 11px;
  line-height: 1.65em;
  margin-bottom: 10px;
}

.c5 {
  font-family: 'ITC Avant Garde Gothic W04','AvantGardeGothicITCW01D 731075',AvantGardeGothicITCW01Dm,Helvetica,sans-serif;
  -webkit-font-smoothing: antialiased;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-size: 13px;
  line-height: 1.65em;
  background: #e5e5e5;
  color: black;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 15px 30px;
  line-height: 1;
  outline: 0;
  -webkit-transition: background-color 0.25s,color 0.25s;
  transition: background-color 0.25s,color 0.25s;
  margin: 10px;
  border: none;
  box-sizing: border-box;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-radius: 2px;
  background: black;
  color: white;
  border-radius: 2px;
  height: 30px;
  width: 80px;
  margin-left: -100px;
  font-family: 'ITC Avant Garde Gothic W04','AvantGardeGothicITCW01D 731075',AvantGardeGothicITCW01Dm,Helvetica,sans-serif;
  -webkit-font-smoothing: antialiased;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-size: 11px;
  line-height: 1.65em;
}

.c5:hover:not(:disabled) {
  cursor: pointer;
  background: #cccccc;
}

.c5:hover:not(:disabled) {
  background: #6e1fff;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    The Daily Editorial Newsletter
  </div>
  <div
    className="c2"
  >
    The latest art-world news and stories in your inbox
  </div>
  <div
    className="c3"
  >
    <input
      className="c4"
      onChange={[Function]}
      placeholder="Enter Your Email..."
      readOnly={false}
      type="email"
      value=""
    />
    <button
      className="c5"
      disabled={false}
      onClick={[Function]}
    >
      <span>
        Subscribe
      </span>
    </button>
  </div>
</div>
`;
